### PR TITLE
[Summer2021 issue#5698] Implement Core Functionalities For Circuit-Breaker 

### PR DESCRIPTION
**Pull request by newJimmyChu**
Please do not create a Pull Request without creating an issue first.

For #5698.

## What is the purpose of the change

Implement core functionalities for Circuit-Breaker & Implement a default circuit-break strategy using our old (TPS Control) code.

## Brief changelog

1. Changed some of the core interfaces to raise the level of abstraction.
2. Finished a default circuit-breaker strategy using our TPS implementation.
3. Refactored saveRuleToLocal() func to adapt new structures.
4. Placed TpsConfig into the TpsRecorder to improve performance.

## Verifying this change

Follow this checklist to help us incorporate your contribution quickly and easily:

* [x] Make sure there is a Github issue filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
* [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
* [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
* [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/alibaba/nacos/tree/master/test).
* [ ] Run `mvn -B clean package apache-rat:check findbugs:findbugs -Dmaven.test.skip=true` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.

### CODE REVIEW

File: core/src/main/java/com/alibaba/nacos/core/remote/circuitbreaker/CircuitBreaker.java [hunk 1]
```diff
@@ -47,22 +56,24 @@
      * @param  pointName entry point name or class name (TODO: can be modified through Nacos console)
      * @return true when the current request is allowed to continue; false if the request breaks the upper limit
      */
```
Comment (shiyiyue1102) [inline core/src/main/java/com/alibaba/nacos/core/remote/circuitbreaker/CircuitBreaker.java:58]: parameter   'connectionId' should be removed

File: core/src/main/java/com/alibaba/nacos/core/remote/circuitbreaker/CircuitBreaker.java [hunk 2]
```diff
@@ -47,22 +56,24 @@
...
+     * @param  connectionId the specific circuit break strategy that client wants to apply for the current point
      *
      * @return true when the current request is allowed to continue; false if the request breaks the upper limit
      */
-    public static boolean applyForStrategy(String pointName, String strategyName) {
-        return true;
+    public boolean applyStrategyForClientIp(String pointName, String connectionId, String clientIp) {
+        // Check monitor keys & check total tps.
+        return currentRule.applyStrategyForClientIp(pointName, connectionId, clientIp);
     }
```
Comment (shiyiyue1102) [inline core/src/main/java/com/alibaba/nacos/core/remote/circuitbreaker/CircuitBreaker.java:77]: parameter clientIp and connectionId should be removed

File: core/src/main/java/com/alibaba/nacos/core/remote/circuitbreaker/CircuitBreaker.java [hunk 3]
```diff
@@ -115,4 +130,61 @@ public  void loadConfig() {
...
+
+    private synchronized void loadRuleFromLocal(String pointName) throws IOException {
+
+        File pointFile = getRuleFile(pointName);
+        if (!pointFile.exists()) {
+            pointFile.createNewFile();
+        }
+        String pointConfigContent = DiskUtils.readFile(pointFile);
+        String monitorKeyContent = DiskUtils.readFile(pointFile);
+
```
Comment (shiyiyue1102) [inline core/src/main/java/com/alibaba/nacos/core/remote/circuitbreaker/CircuitBreaker.java:142]: it seems pointConfigContent and monitorKeyContent are in the same file, and content is consistent?

File: core/src/main/java/com/alibaba/nacos/core/remote/circuitbreaker/CircuitBreaker.java (hunk 1 above)
Comment (newJimmyChu) [reply core/src/main/java/com/alibaba/nacos/core/remote/circuitbreaker/CircuitBreaker.java:58]: sure

File: core/src/main/java/com/alibaba/nacos/core/remote/circuitbreaker/CircuitBreaker.java (hunk 3 above)
Comment (newJimmyChu) [reply core/src/main/java/com/alibaba/nacos/core/remote/circuitbreaker/CircuitBreaker.java:142]: They should be in the separate files. I will fix this in the following commits.

**Pull request merged**